lib/redis_memo/memoize_method.rb

```diff
@@ -81,14 +81,28 @@ def self.extract_dependencies(ref, *method_args, &depends_on)
 
   def self.get_or_extract_dependencies(ref, *method_args, &depends_on)
     if RedisMemo::Cache.local_dependency_cache
-      RedisMemo::Cache.local_dependency_cache[ref] ||= {}
-      RedisMemo::Cache.local_dependency_cache[ref][depends_on] ||= {}
-      RedisMemo::Cache.local_dependency_cache[ref][depends_on][method_args] ||= extract_dependencies(ref, *method_args, &depends_on)
+      RedisMemo::Cache.local_dependency_cache[ref.class] ||= {}
+      RedisMemo::Cache.local_dependency_cache[ref.class][depends_on] ||= {}
+      mapped_args = map_depends_on_method_args(depends_on, method_args)
+      RedisMemo::Cache.local_dependency_cache[ref.class][depends_on][mapped_args] ||= extract_dependencies(ref, *method_args, &depends_on)
     else
       extract_dependencies(ref, *method_args, &depends_on)
     end
   end
 
+  def self.map_depends_on_method_args(depends_on, args)
+    mapped_args = {}
+    unless depends_on.parameters.empty? or args.empty?
+      params = depends_on.parameters.drop(1)
+      params.each_with_index do |param, i|
+        unless param[1] == :_ || param[1].nil?
```

Comment:
When would `param[1].nil?` be true?

Reply:
do we need to worry about
```
do |*args|
end

do |*|
end

do |*args, **kwargs, &blk|
end
```
?

Reply:
> When would param[1].nil? be true?

I think for cases such as 
```
do |*|
end

do |**|
end
```
I tried something like ` do |**_|`, but this didn't seem to work.

I can add more unit tests around the other formats you listed.

Reply:
```
memoize_method :method do |*args, **kwargs, &blk|
end
```
I think this format is a little tricky with how it's currently implemented since we call `dependency.instance_exec(ref, *args, &depends_on)`, where the depends_on block is the block above. I looked into it and not sure how to pass in a block which expects another block to instance_exec? https://stackoverflow.com/questions/15357046/can-i-pass-a-block-which-itself-expect-a-block-to-instance-exec-in-ruby

Reply:
oooh that's a good point. yeah I guess we cannot really support memoizing methods that accept a block.

Reply:
What happens if we have
```ruby
memoize_method :method do |*args, **kwargs|
  depends_on RedisMemo::Memoizable.new(param2: args[0], kwparam: kwargs[:options])
end
```

Reply:
> oooh that's a good point. yeah I guess we cannot really support memoizing methods that accept a block.

I think we can with some changes, but the memoize_method dependency block itself cannot have a block; it would have to be a proc instead (similar to what the stackoverflow post suggests).

Reply:
I wonder if we should always expect that the memoize_method dependency block's first argument is a reference?

e.g. should it be 
```

memoize_method :method do |_, *args, **kwargs|
  depends_on RedisMemo::Memoizable.new(param2: args[0], kwparam: kwargs[:options])
end
```

Reply:
That seems to be a reasonable assumption!

Yet I still wonder what would happen if the `depends_on_args` does not exactly match `depends_on.parameters`:

for example, the `depends_on_args` could contain multiple positional/keyword arguments, but but `depends_on.parameters` might only have one, representing a packed version

Reply:
> > oooh that's a good point. yeah I guess we cannot really support memoizing methods that accept a block.


> I think we can with some changes, but the memoize_method dependency block itself cannot have a block; it would have to be a proc instead (similar to what the stackoverflow post suggests).

Created a separate task for this in clubhouse. Right now though if we try to memoize a method that contains a block, I think that block won't actually get passed to the original method call unless we modify this [line](https://github.com/chanzuckerberg/redis-memo/blob/main/lib/redis_memo/memoize_method.rb#L15) to be `|*args, &blk|`, and pass the blk down?

Reply:
> Created a separate task for this in clubhouse.

Thanks!



> Right now though if we try to memoize a method that contains a block, I think that block won't actually get passed to the original method call unless we modify this [line](https://github.com/chanzuckerberg/redis-memo/blob/main/lib/redis_memo/memoize_method.rb#L15) to be `|*args, &blk|`, and pass the blk down?

Sounds right to me. Though at a high level, memorizing methods with block might not make sense unless all the blocks passed in have a fixed set of dependencies.